Add initial guide for maintaining Software Bill-of-Materials

sethmlarson opened the pull request:
Part of https://github.com/python/cpython/issues/112302
Closes https://github.com/python/devguide/issues/1241

This adds an initial guide for maintaining CPython's SBOM document for core developers and adds a future location for the docs for later additions to CPython's SBOM tooling (e.g. when I implement external/optional dependencies from cpython-source-deps).

<!-- readthedocs-preview cpython-devguide start -->
----
📚 Documentation preview 📚: https://cpython-devguide--1251.org.readthedocs.build/

<!-- readthedocs-preview cpython-devguide end -->

hugovk commented on developer-workflow/sbom.rst:
```diff
@@ -0,0 +1,45 @@
+which can be transformed into other formats if necessary by consumers.
+
+There are multiple sources of third-party dependencies for CPython.
+Some are vendored into the source code of CPython itself (like ``mpdecimal`` vendored at ``Modules/_decimal/libmpdec``)
```
We can link to the CPython repo like this:
```suggestion
Some are vendored into the source code of CPython itself
(like ``mpdecimal`` vendored at :cpy-file:`Modules/_decimal/libmpdec`)
```

hugovk commented on developer-workflow/sbom.rst:
```diff
@@ -0,0 +1,45 @@
+directly, there is a tool for automatically updating the document at ``Tools/build/generate_sbom.py``.
+The recommended workflow is:
+
+* Update the vendored code in-place. Take note of the new version, download URL, and checksum of the downloaded archive.
+* Edit ``Misc/sbom.spdx.json`` to add the new ``versionInfo``, ``downloadLocation``, ``checksums``, and ``externalReferences`` for the corresponding ``package``.
+  Don't update any information in ``files`` and ``relationships`` as this will be generated automatically later.
+* Run ``$ make regen-sbom`` or ``Tools/build/generate_sbom.py``. Ensure that this doesn't fail with validation errors.
+* Run ``$ git diff Misc/sbom.spdx.json`` and check the diff matches the expected changes.
+* Commit the changes to ``Misc/sbom.spdx.json`` along with the update to the dependency code.
```
Numbered list for steps to be followed in order:

```suggestion
1. Update the vendored code in-place. Take note of the new version, download
   URL, and checksum of the downloaded archive.
2. Edit :cpy-file:`Misc/sbom.spdx.json` to add the new ``versionInfo``,
   ``downloadLocation``, ``checksums``, and ``externalReferences`` for the
   corresponding ``package``.
   Don't update any information in ``files`` and ``relationships`` as this will
   be generated automatically later.
3. Run ``make regen-sbom`` or ``python Tools/build/generate_sbom.py``. Ensure
   that this doesn't fail with validation errors.
4. Run ``git diff Misc/sbom.spdx.json`` and check the diff matches the expected
   changes.
5. Commit the changes to :cpy-file:`Misc/sbom.spdx.json` along with the update
   to the dependency code.
```

hugovk commented on developer-workflow/sbom.rst:
```diff
@@ -0,0 +1,45 @@
+Software Bill-of-Materials (abbreviated as "SBOM") is a document for sharing information about
+software and how it's been composed. This format is used most often in the security space for
+checking software and its dependencies for vulnerabilities using vulnerability databases like
+`CVE <https://www.cve.org/>`_ and `OSV <https://osv.dev/>`_. The SBOM format that the CPython project uses is `SPDX <https://spdx.github.io/spdx-spec/v2.3/>`_
```
Please could you wrap this line and the others to max 80 chars?

hugovk commented on developer-workflow/sbom.rst:
```diff
@@ -0,0 +1,45 @@
+* License of the project as an `SPDX License Expression <https://spdx.org/licenses/>`_
+* Software identifiers that match values in vulnerability databases (`CPE <https://nvd.nist.gov/products/cpe>`_ and `Package URLs <https://github.com/package-url/purl-spec/blob/master/PURL-SPECIFICATION.rst>`_ or "PURLs")
+
+Add the information into a new entry in ``packages`` in the file ``Misc/sbom.spdx.json``.
+If a new license ID is to be used, add the license expression to ``ALLOWED_LICENSE_EXPRESSIONS``
+in the SBOM generation tool.
```
```suggestion
1. Add the information into a new entry in ``packages`` in the file
   :cpy-file:`Misc/sbom.spdx.json`.
2. If a new license ID is to be used, add the license expression to
   ``ALLOWED_LICENSE_EXPRESSIONS`` in the SBOM generation tool.
```

hugovk commented on developer-workflow/sbom.rst:
```diff
@@ -0,0 +1,45 @@
+If a new license ID is to be used, add the license expression to ``ALLOWED_LICENSE_EXPRESSIONS``
+in the SBOM generation tool.
+
+When removing a dependency, remove the entry from the ``Misc/sbom.spdx.json`` under the ``packages`` field
+and then remove the corresponding entry in ``Tools/build/generate_sbom.py`` and then run ``make regen-sbom``.
```
```suggestion
When removing a dependency:

1. Remove the entry from the :cpy-file:`Misc/sbom.spdx.json` under the
   ``packages`` field.
2. Remove the corresponding entry in :cpy-file:`Tools/build/generate_sbom.py`.
3. Run ``make regen-sbom``.
```

hugovk commented on developer-workflow/sbom.rst:
```diff
@@ -0,0 +1,45 @@
+Updating a dependency
+---------------------
+
+The SBOM for CPython's bundled dependencies is kept at ``Misc/sbom.spdx.json``. Instead of updating this document
+directly, there is a tool for automatically updating the document at ``Tools/build/generate_sbom.py``.
```
```suggestion
The SBOM for CPython's bundled dependencies is kept at
:cpy-file:`Misc/sbom.spdx.json`. Instead of updating this document directly,
there is a tool for automatically updating the document at
:cpy-file:`Tools/build/generate_sbom.py`.
```

hugovk reviewed (commented):
Thanks for adding this!

sethmlarson commented:
Thanks for the review @hugovk, I believe I've incorporated all of your suggestions :)

hugovk commented on developer-workflow/sbom.rst:
```diff
@@ -0,0 +1,62 @@
+
+When adding a dependency it's important to have the following information:
+
+1. Name, version, and download URL of the project
+2. License of the project as an `SPDX License Expression <https://spdx.org/licenses/>`_
+3. Software identifiers that match values in vulnerability databases
+   (`CPE <https://nvd.nist.gov/products/cpe>`_ and
+   `Package URLs <https://github.com/package-url/purl-spec/blob/master/PURL-SPECIFICATION.rst>`_
+   or "PURLs")
```
Let's keep these as bulleted list. Eeasoning is they're not steps to follow, and could be in any order.

* https://developers.google.com/tech-writing/one/lists-and-tables
* https://developers.google.com/style/lists

sethmlarson replied:
Ah sorry, this was my brain in auto-pilot updating the wrong list. I'll fix this.

hugovk commented on developer-workflow/sbom.rst:
```diff
@@ -0,0 +1,62 @@
+   `Package URLs <https://github.com/package-url/purl-spec/blob/master/PURL-SPECIFICATION.rst>`_
+   or "PURLs")
+
+Add the information into a new entry in ``packages`` in the file
+:cpy-file:`Misc/sbom.spdx.json`. If a new license ID is to be used, add the
+license expression to ``ALLOWED_LICENSE_EXPRESSIONS`` in the SBOM generation tool.
```
But we could put these as a couple of numbered steps. I think it helps, as we had a numbered list of steps above for updating, so the reader can quickly see, aha, here's more steps to follow.

hugovk commented on developer-workflow/sbom.rst:
```diff
@@ -0,0 +1,62 @@
+:cpy-file:`Misc/sbom.spdx.json`. If a new license ID is to be used, add the
+license expression to ``ALLOWED_LICENSE_EXPRESSIONS`` in the SBOM generation tool.
+
+When removing a dependency, remove the entry from the :cpy-file:`Misc/sbom.spdx.json`
+under the ``packages`` field and then remove the corresponding entry in
+:cpy-file:`Tools/build/generate_sbom.py` and then run ``make regen-sbom``.
```
And likewise could have three numbered steps here. (See also my earlier suggestions for what I meant.)

Google calls having several things in a sentence like this an "embedded list", and it's clearer to expand them. Especially as we often don't read text on screen, but scan.

sethmlarson commented:
Thanks for the review @hugovk, I've done the following:

* Flattened the embedded lists as you recommended.
* Switched the numbered list of items back to unnumbered
* Split the "Remove" section to be separate
* Added a lot more detail into describing which SPDX SBOM fields to use for information.

hugovk commented on developer-workflow/sbom.rst:
```diff
@@ -0,0 +1,109 @@
+   update the embedded version within URLs and project identifiers.
+   Don't update any information in ``files`` and ``relationships`` as this will
+   be generated automatically by the SBOM tool.
+4. Run ``$ make regen-sbom`` or :cpy-file:`Tools/build/generate_sbom.py`.
+   Ensure that this doesn't fail with validation errors.
+5. Run ``$ git diff Misc/sbom.spdx.json`` and check the diff matches the
```
Let's remove the dollars from inline commands:

```suggestion
4. Run ``make regen-sbom`` or `` python Tools/build/generate_sbom.py``.
   Ensure that this doesn't fail with validation errors.
5. Run ``git diff Misc/sbom.spdx.json`` and check the diff matches the
```

hugovk commented on developer-workflow/sbom.rst:
```diff
@@ -0,0 +1,109 @@
+  (`CPE <https://nvd.nist.gov/products/cpe>`_ and
+  `Package URLs <https://github.com/package-url/purl-spec/blob/master/PURL-SPECIFICATION.rst>`_
+  or "PURLs")
+* Path(s) to include and exclude in the CPython source tree corresponding to this dependency.
```
```suggestion
* Paths to include and exclude in the CPython source tree corresponding to this dependency
```

hugovk commented on developer-workflow/sbom.rst:
```diff
@@ -0,0 +1,109 @@
+   ``SPDXRef-PACKAGE-*`` prefix) in :cpy-file:`Tools/build/generate_sbom.py`.
+4. Run the tool with ``make regen-sbom`` or ``python Tools/build/generate_sbom.py``.
+   Ensure that the tool doesn't fail with any validation errors.
+5. Compare the changes to :cpy-file:`Misc/sbom.spdx.json` with ``$ git diff``, check
```
```suggestion
5. Compare the changes to :cpy-file:`Misc/sbom.spdx.json` with ``git diff``, check
```

hugovk commented on developer-workflow/sbom.rst:
```diff
@@ -0,0 +1,109 @@
+2. Remove the corresponding ``PackageFiles`` entry in :cpy-file:`Tools/build/generate_sbom.py`
+3. Run the tool with ``make regen-sbom`` or ``python Tools/build/generate_sbom.py``.
+   Ensure that the tool doesn't fail with any validation errors.
+4. Compare the changes to :cpy-file:`Misc/sbom.spdx.json` with ``$ git diff``, check
```
```suggestion
4. Compare the changes to :cpy-file:`Misc/sbom.spdx.json` with ``git diff``, check
```

ezio-melotti replied:
I was just about to suggest the same, but you beat me to it :)

hugovk reviewed (commented):
Thanks, looks good! Just some little nits!

ezio-melotti commented on developer-workflow/sbom.rst:
```diff
@@ -0,0 +1,109 @@
+   auto-format your manually written JSON. The fields to fill out include:
+
+   * ``name`` for the project name.
+   * ``SPDXID`` which will be ``"SPDXRef-PACKAGE-{name}"``
```
```suggestion
   * ``SPDXID`` which will be ``"SPDXRef-PACKAGE-{name}"``.
```

sethmlarson commented:
@hugovk Done! Thanks for teaching me about that GitHub feature btw, I hadn't used it before since it always seemed to be greyed out :)

hugovk commented on developer-workflow/sbom.rst:
```diff
@@ -0,0 +1,109 @@
+   update the embedded version within URLs and project identifiers.
+   Don't update any information in ``files`` and ``relationships`` as this will
+   be generated automatically by the SBOM tool.
+4. Run ``make regen-sbom`` or `` python Tools/build/generate_sbom.py``.
```
Sorry I broke the linter:

```suggestion
4. Run ``make regen-sbom`` or ``python Tools/build/generate_sbom.py``.
```

hugovk reviewed (approved):
Thank you!

We can link to this page on the CI when `make regen-sbom` fails.

sethmlarson commented:
I'm happy to merge this whenever, I'll take it as an action to link to the new page once available in the `regen-sbom` CI job too.

hugovk commented:
Let's merge now, we can make adjustments whenever we need to. Thanks again!
